From 229dd40ffe6e47833e48a254779e536852542f18 Mon Sep 17 00:00:00 2001
From: mj <mj@maxlinear.com>
Date: Tue, 10 Jun 2025 22:12:05 +0530
Subject: [PATCH] WLANRTSYS-88676 ZWDFS - switch to second best channel after
 radar detect
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

ZWDFS new design:
Current implementation:

     When radar is hit, we will move to non-DFS channel immediately.
If the DFS channel is the next best channel, then CAC is triggered for that channel via ZWDFS antenna.
Once the CAC is cleared, then the DFS channel configured as an operating channel.
So, moving to DFS channel immediately after radar hit won’t happen, we have to wait for the CAC to finish.
Until then we will be operating in the non-DFS channel.

Implementation needed:

      The DFS channel monitoring should happen continuously in the background for the next best DFS channel to clear the CAC. update the DFS channel state (usable or available or unavailable) During radar hit, do not switch to non-DFS immediately, switch to the second best available DFS channel immediatly.
---
 src/ap/dfs.c                 |  4 +++
 src/ap/drv_callbacks.c       |  4 +++
 src/ap/mxl_acs_proprietary.c | 59 ++++++++++++++++++++++++++++--------
 src/ap/mxl_dfs_common.h      |  3 ++
 src/ap/mxl_drv_callbacks.c   |  1 +
 src/ap/mxl_drv_callbacks.h   |  3 +-
 src/ap/mxl_hostapd.c         |  7 +++++
 src/ap/mxl_hostapd.h         | 16 +++++++++-
 8 files changed, 82 insertions(+), 15 deletions(-)

diff --git a/src/ap/dfs.c b/src/ap/dfs.c
index 976869567..0e23ea117 100644
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1639,6 +1639,10 @@ int hostapd_dfs_radar_detected(struct hostapd_iface *iface, int freq,
 #ifdef CONFIG_VENDOR_MXL
 	mxl_hostapd_dfs_file_radar(iface);
 
+	if (mxl_switch_second_best_channel(iface)) {
+		return 0;
+	}
+
 	if (mxl_hostapd_handle_background_radar(iface, freq, cf1, chan_width))
 		return 0;
 
diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 66b3bc366..44613f6db 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -2188,6 +2188,10 @@ static void hostapd_event_dfs_cac_finished(struct hostapd_data *hapd,
 	hostapd_dfs_complete_cac(hapd->iface, 1, radar->freq, radar->ht_enabled,
 				 radar->chan_offset, radar->chan_width,
 				 radar->cf1, radar->cf2);
+#ifdef CONFIG_VENDOR_MXL
+	/* csa_iface reset if we do CAC manually at run time. So check csa pointer to do CAC with second best channel */
+	mxl_check_csa_and_get_second_best_channel(hapd);
+#endif /* CONFIG_VENDOR_MXL */
 }
 
 
diff --git a/src/ap/mxl_acs_proprietary.c b/src/ap/mxl_acs_proprietary.c
index 97cdf337d..70c4a99db 100644
--- a/src/ap/mxl_acs_proprietary.c
+++ b/src/ap/mxl_acs_proprietary.c
@@ -2454,14 +2454,6 @@ typedef struct {
 #define MAX(x,y) ((x)>(y)?(x):(y))
 #endif
 
-static int get_secondary_offset(int primary, int secondary)
-{
-	if (secondary)
-		return ((secondary > primary) ? 1 : -1);
-	else
-		return 0;
-}
-
 int hostapd_csa_in_progress(struct hostapd_iface *iface);
 int hostapd_dfs_start_channel_switch_cac(struct hostapd_iface *iface);
 
@@ -2702,7 +2694,7 @@ static bool mxl_acs_set_new_chan_if_ok(struct hostapd_iface *iface, int min_rank
 	/* set the new channel */
 	iface->conf->channel = candidates[min_rank_cand_idx].primary;
 	iface->freq = hostapd_hw_get_freq(iface->bss[0], iface->conf->channel);
-	iface->conf->secondary_channel = get_secondary_offset(candidates[min_rank_cand_idx].primary, candidates[min_rank_cand_idx].secondary);
+	iface->conf->secondary_channel = mxl_get_secondary_offset(candidates[min_rank_cand_idx].primary, candidates[min_rank_cand_idx].secondary);
 	iface->mxl_iface.selected_candidate = min_rank_cand_idx;
 	hostapd_set_oper_chwidth(iface->conf, mxl_get_oper_width(candidates[min_rank_cand_idx].width));
 	acs_smart_adjust_center_freq(iface, candidates[min_rank_cand_idx].chan);
@@ -2762,7 +2754,7 @@ static bool mxl_acs_check_new_chan_is_ok(struct hostapd_iface *iface, int min_ra
 				/* set the previous channel */
 				iface->conf->channel = candidates[iface->mxl_iface.selected_candidate].primary;
 				iface->freq = hostapd_hw_get_freq(iface->bss[0], iface->conf->channel);
-				iface->conf->secondary_channel = get_secondary_offset(candidates[iface->mxl_iface.selected_candidate].primary, candidates[iface->mxl_iface.selected_candidate].secondary);
+				iface->conf->secondary_channel = mxl_get_secondary_offset(candidates[iface->mxl_iface.selected_candidate].primary, candidates[iface->mxl_iface.selected_candidate].secondary);
 				hostapd_set_oper_chwidth(iface->conf, mxl_get_oper_width(candidates[iface->mxl_iface.selected_candidate].width));
 				acs_smart_adjust_center_freq(iface, candidates[iface->mxl_iface.selected_candidate].chan);
 			}
@@ -3563,12 +3555,34 @@ int mxl_acs_recalc_ranks_and_set_chan (struct hostapd_iface *iface, int switch_r
 		int cg_idx = CALIB_20;
 		int cwi_rank, chan_load_rank, tx_power_rank, num_bss_rank = 0;
 		int rank_util_primary = 0;
+		bool acs_cand_skip = false;
+		int first_rank_freq = hostapd_hw_get_freq(iface->bss[0], iface->conf->channel);
+		int first_rank_width = mxl_hostapd_get_num_width(hostapd_get_oper_chwidth(iface->conf), iface->conf->secondary_channel);
 
 		wpa_printf(MSG_DEBUG, "ACS: recalculate candidate idx %d", i);
 
+		if (iface->mxl_iface.sec_best_ch.flag) {
+			acs_print_cand_no_file(fp, iface, i);
+			acs_cand_skip = true;
+		}
+
 		acs_print_info_header(MSG_DEBUG);
 		acs_print_cand_no(MSG_DEBUG, iface, i);
 
+		if (iface->mxl_iface.sec_best_ch.ch_switch_done) {
+			first_rank_freq = iface->mxl_iface.sec_best_ch.freq;
+			first_rank_width = iface->mxl_iface.sec_best_ch.width;
+			iface->mxl_iface.sec_best_ch.ch_switch_done = false;
+			wpa_printf(MSG_DEBUG, "ACS: ch_switch_done sec_best_ch freq : %d sec_best_ch width : %d", iface->mxl_iface.sec_best_ch.freq, iface->mxl_iface.sec_best_ch.width);
+		}
+
+		if (iface->mxl_iface.sec_best_ch.first_rank_idx && iface->mxl_iface.sec_best_ch.flag) {
+			if (channels_overlap(first_rank_freq, first_rank_width, candidates[i].freq, candidates[i].width)) {
+				wpa_printf(MSG_ERROR, "channel overlapped  first freq %d, : candidates[i].freq %d", candidates[iface->mxl_iface.sec_best_ch.first_rank_idx].freq, candidates[i].freq);
+				continue;
+			}
+		}
+
 		candidates[i].exclude_reason = acs_to_exclude_candidate(iface, i, &now);
 
 		if (iface->mxl_iface.sec_best_ch.flag) {
@@ -3711,8 +3725,9 @@ int mxl_acs_recalc_ranks_and_set_chan (struct hostapd_iface *iface, int switch_r
 
 		wpa_printf(MSG_DEBUG, "ACS: candidate idx %d rank %d, bw_rank %d, primary util %d, util %d",
 			   i, candidates[i].rank, candidates[i].bw_comp_rank, candidates[i].rank_util_primary, candidates[i].bw_comp_util);
-
-		acs_print_cand_no_file(fp, iface, i);
+		/* Avoid duplicate print in the acs table */
+		if (!acs_cand_skip)
+			acs_print_cand_no_file(fp, iface, i);
 
 		if (candidates[i].exclude_reason != ACS_NO_EXCLUSION &&
 		    candidates[i].exclude_reason != ACS_EXCLUDE_LOW_PRIMARY_UTIL)
@@ -3821,6 +3836,18 @@ int mxl_acs_recalc_ranks_and_set_chan (struct hostapd_iface *iface, int switch_r
 
 set_chan:
 		wpa_printf(MSG_DEBUG, "ACS: exit recalculate, selected prio %d min_rank_cand_idx %d", prio, min_rank_cand_idx);
+		if (iface->mxl_iface.sec_best_ch.flag) {
+			wpa_printf(MSG_DEBUG, "ACS: prio %d second rank_cand_idx %d,  chan : %d", prio, min_rank_cand_idx, candidates[min_rank_cand_idx].freq);
+			iface->mxl_iface.sec_best_ch.second_rank_idx = min_rank_cand_idx;
+			iface->mxl_iface.sec_best_ch.channel = candidates[min_rank_cand_idx].chan;
+			iface->mxl_iface.sec_best_ch.freq = candidates[min_rank_cand_idx].freq;
+			iface->mxl_iface.sec_best_ch.width = candidates[min_rank_cand_idx].width;
+			iface->mxl_iface.sec_best_ch.primary = candidates[min_rank_cand_idx].primary;
+			iface->mxl_iface.sec_best_ch.secondary = candidates[min_rank_cand_idx].secondary;
+			return 0;
+		}
+		iface->mxl_iface.sec_best_ch.first_rank_idx = min_rank_cand_idx;
+
 		res = mxl_acs_check_new_chan_is_ok(iface, min_rank_cand_idx, switch_reason, fp_hist);
 #ifdef CONFIG_AFC
 		/* In SP/ISP, we will rank the best candidate + valid punctured pattern
@@ -3870,7 +3897,7 @@ set_chan:
 	/* if we are still here - use the fallback channel */
 	iface->conf->channel = iface->conf->mxl_conf.acs_fallback_chan.primary;
 	iface->freq = hostapd_hw_get_freq(iface->bss[0], iface->conf->channel);
-	iface->conf->secondary_channel = get_secondary_offset(iface->conf->mxl_conf.acs_fallback_chan.primary, iface->conf->mxl_conf.acs_fallback_chan.secondary);
+	iface->conf->secondary_channel = mxl_get_secondary_offset(iface->conf->mxl_conf.acs_fallback_chan.primary, iface->conf->mxl_conf.acs_fallback_chan.secondary);
 	hostapd_set_oper_chwidth(iface->conf, mxl_get_oper_width(iface->conf->mxl_conf.acs_fallback_chan.width));
 	acs_smart_adjust_center_freq(iface, 0);
 	wpa_printf(MSG_DEBUG, "ACS: exit recalculate, using fallback channel");
@@ -4608,6 +4635,10 @@ void mxl_smart_acs_scan_complete(struct hostapd_iface *iface)
 	}
 
 	mxl_acs_smart_process_bsses(iface, scan_res);
+	if (WAVE_FREQ_IS_5G(iface->freq)) {
+		iface->mxl_iface.sec_best_ch.flag = true;
+		mxl_acs_recalc_ranks_and_set_chan(iface, SWR_INITIAL);
+	}
 
 	/* After 2.4 and 5 RNR Processing call 6G ACS Scan Request with RNR Data */
 	if (!(is_6ghz_op_class(iface->conf->op_class))) {
@@ -4821,6 +4852,8 @@ void mxl_acs_init_defaults(struct hostapd_iface *iface)
 		candidates[i].cwi_noise[SEC_MAX_80] = 0;
 		candidates[i].cwi_noise[SEC_MAX_160] = 0;
 	}
+	/* reset second best channel params */
+	memset(&iface->mxl_iface.sec_best_ch, 0, sizeof(iface->mxl_iface.sec_best_ch));
 }
 
 /* Gets secondary channel offset and VHT operational center frequency segment 0 index
diff --git a/src/ap/mxl_dfs_common.h b/src/ap/mxl_dfs_common.h
index 8e08046cc..5112147bc 100644
--- a/src/ap/mxl_dfs_common.h
+++ b/src/ap/mxl_dfs_common.h
@@ -26,6 +26,7 @@ enum mxl_hostapd_cac_finished_stat {
 	HOSTAPD_CAC_STAT_PAUSED  = 2
 };
 
+int mxl_get_secondary_offset(int primary, int secondary);
 int mxl_hostapd_drv_set_antenna(struct hostapd_iface *iface, u32 tx, u32 rx);
 void mxl_hostapd_event_dfs_cac_paused(struct hostapd_data *hapd, struct dfs_event *radar);
 struct hostapd_channel_data *
@@ -84,5 +85,7 @@ mxl_dfs_get_channel_dynamic_failsafe(struct hostapd_iface *iface, int *secondary
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
 void mxl_dfs_cut_cw_by_half(struct hostapd_iface *iface, bool backup);
 void mxl_dfs_restore_original_cw(struct hostapd_iface *iface);
+bool mxl_validate_second_best_chan(struct hostapd_iface *iface);
+bool mxl_switch_second_best_channel(struct hostapd_iface *iface);
 
 #endif /* MXL_DFS_COMMON_H */
diff --git a/src/ap/mxl_drv_callbacks.c b/src/ap/mxl_drv_callbacks.c
index 17867c1e0..5eed7661f 100644
--- a/src/ap/mxl_drv_callbacks.c
+++ b/src/ap/mxl_drv_callbacks.c
@@ -23,6 +23,7 @@
 #include "wpa_auth.h"
 #include "ap_drv_ops.h"
 #include "beacon.h"
+#include "mxl_hostapd.h"
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 void mxl_hostapd_notif_disassoc(struct hostapd_data *hapd, const u8 *addr)
diff --git a/src/ap/mxl_drv_callbacks.h b/src/ap/mxl_drv_callbacks.h
index 4778c616d..d27fdbd03 100644
--- a/src/ap/mxl_drv_callbacks.h
+++ b/src/ap/mxl_drv_callbacks.h
@@ -26,5 +26,6 @@ void mxl_hostapd_event_eapol_rx(struct hostapd_data *hapd, const u8 *src,
 				enum frame_encryption encrypted,
 				int link_id,
 				u16 vlan_id);
-
+void mxl_check_ch_switch_status_and_get_second_best_channel(int finished, struct hostapd_data *hapd);
+void mxl_check_csa_and_get_second_best_channel(struct hostapd_data *hapd);
 #endif /* MXL_DRV_CALLBACKS_H */
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 58474c75c..16fdc78c9 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -38,6 +38,7 @@
 #endif /* CONFIG_AFC */
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
 #include "mxl_acs.h"
+#include "mxl_acs_proprietary.h"
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
 
 #define COC_POWER_1x1_STATE 1
@@ -1783,6 +1784,12 @@ void mxl_hostapd_setup_interface_post(struct hostapd_iface *iface)
 	struct mxl_hostapd_config *mxl_conf = &hapd->iconf->mxl_conf;
 	int ap_retry_limit_cfg[3];
 
+#ifdef CONFIG_VENDOR_MXL_PROPRIETARY
+	if (WAVE_FREQ_IS_5G(iface->freq) && !(iface->mxl_iface.sec_best_ch.bg_cac)) {
+		iface->mxl_iface.sec_best_ch.flag = true;
+		mxl_get_second_best_channel(iface);
+	}
+#endif
 	mxl_acs_set_bgscan_interval(hapd);
 
 	/* Set AP retry limit only if user configured this */
diff --git a/src/ap/mxl_hostapd.h b/src/ap/mxl_hostapd.h
index ce66c7dfd..51b6759db 100644
--- a/src/ap/mxl_hostapd.h
+++ b/src/ap/mxl_hostapd.h
@@ -95,6 +95,19 @@ struct hostapd_failsafe_freq_params {
 	u32 tx_ant, rx_ant;
 };
 
+struct acs_second_best_channel_params {
+	u8 first_rank_idx;
+	u8 second_rank_idx;
+	bool ch_switch_done;
+	bool bg_cac;
+	bool flag;
+	u32 channel;
+	u32 freq;
+	u32 primary;
+	u32 secondary;
+	int width;
+};
+
 struct mxl_hostapd_iface {
 	u32 sta_aid_link[MXL_AID_WORDS];
 	s8 stas_free_reason; /* the reason for all stations remove */
@@ -102,7 +115,7 @@ struct mxl_hostapd_iface {
 	unsigned int acs_update_in_progress:1;
 	unsigned int cac_required:1;
 	int in_scan;
-
+	struct acs_second_best_channel_params sec_best_ch;
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
 #ifdef CONFIG_ACS
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
@@ -273,6 +286,7 @@ bool mxl_hostapd_ieee80211n_check_40mhz(struct hostapd_iface *iface);
 int mxl_hostapd_restart_iface(struct hostapd_iface *iface);
 void mxl_dynamic_wmm_param_init(struct hostapd_data *hapd);
 void mxl_hostapd_setup_interface_pre_wmm(struct hostapd_iface *iface, struct hostapd_data *hapd);
+void mxl_get_second_best_channel(struct hostapd_iface *iface);
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
 
 int mxl_hostapd_get_num_width(enum oper_chan_width width, int secondary);
-- 
2.43.0

